libp2p/transport/tcp/tcp.py

```diff
@@ -91,5 +100,52 @@ def create_listener(self, handler_function: THandler) -> TCPListener:
         return TCPListener(handler_function)
 
 
-def _multiaddr_from_socket(socket: socket) -> Multiaddr:
-    return Multiaddr("/ip4/%s/tcp/%s" % socket.getsockname())
+def _ip4_or_6_from_multiaddr(maddr: Multiaddr) -> Optional[str]:
+    if P_IP4 in maddr.protocols():
+        return maddr.value_for_protocol(P_IP4)
+    elif P_IP6 in maddr.protocols():
+        return maddr.value_for_protocol(P_IP6)
+    else:
+        return None
+
+
+def _multiaddr_from_socket(sock: socket.socket) -> Multiaddr:
+    # Reference: http://man7.org/linux/man-pages/man2/socket.2.html#DESCRIPTION
+    # todo: move this to more generic libp2p.transport helper function
+
+    # Reference: https://stackoverflow.com/questions/5815675/what-is-sock-dgram-and-sock-stream
+    # Selects first protocol in sequence if bitwise AND matches, else None
+    t_proto = next(
+        (
+            v
+            for k, v in {
+                socket.SOCK_STREAM: p_code(P_TCP).name,
+                socket.SOCK_DGRAM: p_code(P_UDP).name,
+            }.items()
+            if k & sock.type != 0
+        ),
+        None,
+    )
+
+    if t_proto is None:
+        raise NotImplementedError(
+            f"Cannot convert socket to multiaddr, socket type is of {sock.type}"
+        )
+
+    # Reference: https://docs.python.org/3/library/socket.html#socket-families
+    if sock.family == socket.AF_INET:
+        # ipv4: (host, port)
+        addr, port = sock.getsockname()
+        ip = p_code(P_IP4).name
+
+    elif sock.family == socket.AF_INET6:
+        # ipv6: (host, port, flowinfo, scopeid)
+        addr, port = sock.getsockname()[:2]
+        ip = p_code(P_IP6).name
+
+    else:
+        raise NotImplementedError(
+            f"Cannot convert socket to multiaddr, socket family is of {sock.family}"
+        )
+
+    return Multiaddr(f"/{ip}/{addr}/{t_proto}/{port}")
```

Comment:
I started changing `Multiaddr("/ip4/%s/tcp/%s" % socket.getsockname())` till I realized it needed an actual working function, so I made one.

Reply:
Typically we would request that this type of substantive change to functionality be submitted separately from the other changes.  Do you think you could extract this refactor into it's own PR that we'll review and merge separately and this branch can be based on top of that other branch?

Reply:
Yes, I'll do that.